proposals/0000-decidable-instances.rst

```diff
@@ -0,0 +1,143 @@
+
+Motivation
+------------
+
+Consider
+
+:: 
+
+ class Monad m => MonadState s m | m -> s 
+
+
+In this scenario, instances like
+
+::
+
+ instance MonadState s m => MonadState s (ReaderT e m)
+
+require ``UndecidableInstances`` today, but are in theory perfectly decidable. The type ``s`` isn't being refined
+by structural recursion, but the type ``ReaderT e m`` that determines it is. Today to compile this instance in the
+``mtl`` we have to turn on the ``UndecidableInstances`` check, and risk accidentally writing instances that really can
+send execution into a tail-spin, and then reason case by case through each instance to see whether or not the instances
+in question are truly undecidable.
+
+Proposed Change Specification
+-----------------------------
+
+The proposal is to add an extension ``DecidableInstances`` that serves as a middle ground between the existing rules
+for structural recursion in instance handling and the free-forall offered by ``UndecidableInstances``.
+
+Under ``DecidableInstances`` the decidability check is relaxed so that instead of requiring structural recursion on
+every parameter of a multi-parameter type class individually, each parameter merely requires that it is functionally
+dependent upon a set of properly structurally-recursive parameters.
+
+To introduce some vocabulary, take ``<=`` to be the relation "is a subset of" then the set of all super-sets of a
+particular set ``p`` is a `principal ideal <https://en.wikipedia.org/wiki/Ideal_(order_theory)#Basic_definitions>`_
+``↑p = { x | p <= x }`` and the set of all subsets of a particular set ``q`` is the
+`principal filter <https://en.wikipedia.org/wiki/Filter_(mathematics)#Filter_on_a_set>`_ ``↓q = { x | x <= q }``.
+
+It will be convenient to group the parameters in a functional dependency. Since each functional dependency logically
+implies all of the weaker functional dependencies obtained by taking a super-set of the inputs and/or sub-set of the
+outputs, let's denote each functional dependency ``a b c .. -> d e f ..`` as ``↑p -> ↓q`` for the sets
+``p = {a,b,c..}`` and ``q = {d,e,f..}``. Note: ``p`` and ``q`` are sets of parameters, and while ``↑p`` and ``↓q`` are
+logically sets of sets, they are representable simply by storing their principal element ``p`` or ``q`` respective.
+
+Given an instance, we can compute the set of class parameters for which the instance performs a simple structural
+recursion. Let's call this set the set ``P`` of known-good structurally-recursive parameters.
+
+Given a set ``P`` of known-good structurally-recursive parameters, and a functional dependency ``↑p -> ↓q``, if ``p``
+is a subset of ``P``, then all parameters in the principal ``q`` should be added to the set of known-good
+structurally-recursive parameters.
+
+Moreover, each functional dependency can only successfuly "fire" once, after which it can be removed from consideration,
```

Comment:
For purposes of understanding the algorithm description that follows, would this be a good place to define *firing* a fundep?

Reply:
I should modify this to be more careful to use common terminology on that line and the previous line, to be clearer.